### PR TITLE
Refactor node attributes

### PR DESCRIPTION
**Pull request by JPXKQX**
This PR introduces the `NamedNodesAttributes(nn.Module)` class, designed to enhance the management of node attributes within graph neural networks. Key features include:

- **Node attribute tensor**: The class provides a tensor that combines coordinates with trainable parameters for each set of nodes in the graph.
- **Grid size**: It retains the `grid_size` attribute, indicating the total number of nodes.
- **Channel dimension**: It tracks `in_channels_src/dst`, which represents the dimensionality of the node attribute tensor.

The main improvement is the ability to handle an arbitrary number of nodes. This development will facilitate extending AnemoiModelEncProcDec to other use cases, such as hierarchical models or multi-encoder/decoder.

- Tested on 1, 2 & 4 GPUs (with model sharding & transformer)
- Tested on 1 GPU: gnn, graphtransformer & transformer
- Tested on 1 GPU: lam & stretched

### CODE REVIEW

File: src/anemoi/models/layers/graph.py [hunk 1]
```diff
@@ -36,8 +37,47 @@ def __init__(self, tensor_size: int, trainable_size: int) -> None:
...
         return torch.cat(
             latent,
             dim=-1,  # feature dimension
         )
+
+
+class NamedNodesAttributes(torch.nn.Module):
+    """Named Node Attributes Module."""
+
+    def __init__(self, num_trainable_params: int, graph_data: HeteroData) -> None:
```
Comment (HCookie) [inline src/anemoi/models/layers/graph.py:50]: It may be useful to type hint the properties you are assigning to within the init

**Review by HCookie — APPROVED** [review]
LGTM
Maybe consider adding some property typehinting

**Review by HCookie — APPROVED** [review]

File: src/anemoi/models/layers/graph.py [hunk 2]
```diff
@@ -36,8 +37,62 @@ def __init__(self, tensor_size: int, trainable_size: int) -> None:
...
         if self.trainable is not None:
-            latent.append(einops.repeat(self.trainable, "e f -> (repeat e) f", repeat=batch_size))
+            latent.append(einops.repeat(self.trainable.to(x.device), "e f -> (repeat e) f", repeat=batch_size))
         return torch.cat(
             latent,
             dim=-1,  # feature dimension
         )
+
+
+class NamedNodesAttributes(torch.nn.Module):
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/graph.py]: nit: Could you make this `nn.Module` or adjust `TrainableTensor` to be consistent?

File: src/anemoi/models/layers/graph.py [hunk 3]
```diff
@@ -36,8 +37,62 @@ def __init__(self, tensor_size: int, trainable_size: int) -> None:
...
+        self.define_fixed_attributes(graph_data, num_trainable_params)
+
+        self.trainable_tensors = nn.ModuleDict()
+        for nodes_name, nodes in graph_data.node_items():
+            self.register_coordinates(nodes_name, nodes.x)
+            self.register_tensor(nodes_name, num_trainable_params)
+
+    def define_fixed_attributes(self, graph_data: HeteroData, num_trainable_params: int) -> None:
+        """Define fixed attributes."""
+        nodes_names = list(graph_data.node_types)
```
Comment (JesperDramsch) [inline src/anemoi/models/layers/graph.py:80]: I assume this was supposed to be an attribute as per line `62`?

File: src/anemoi/models/models/encoder_processor_decoder.py [hunk 4]
```diff
@@ -56,42 +56,33 @@ def __init__(
...
         self.encoder = instantiate(
             model_config.model.encoder,
             in_channels_src=input_dim,
-            in_channels_dst=self.latlons_hidden.shape[1] + self.trainable_hidden_size,
+            in_channels_dst=self.node_attributes.attr_ndims[self._graph_name_hidden],
             hidden_dim=self.num_channels,
             sub_graph=self._graph_data[(self._graph_name_data, "to", self._graph_name_hidden)],
-            src_grid_size=self._data_grid_size,
-            dst_grid_size=self._hidden_grid_size,
+            src_grid_size=self.node_attributes.num_nodes[self._graph_name_data],
```
Comment (JesperDramsch) [inline src/anemoi/models/models/encoder_processor_decoder.py:75]: Is this in anticipation for multi-encoder changes? 
I think we lose some readability this way, which only makes sense if this in preparation for something, I would say.

**Review by JesperDramsch — COMMENTED** [review]
Thanks for tagging me in. Looks very nice. I have some questions, but I think generally this is also good to go.

File: src/anemoi/models/models/encoder_processor_decoder.py (hunk 4 above)
Comment (JPXKQX) [reply src/anemoi/models/models/encoder_processor_decoder.py:75]: Yes! and for hierarchical processor (#37)

Comment (JPXKQX) [reply src/anemoi/models/models/encoder_processor_decoder.py:75]: Another option to set the src_grid_size would be:
```
 src_grid_size=self.graph_data[self._graph_name_data].num_nodes,
```
not sure which one is more readable.

**Review by JesperDramsch — APPROVED** [review]

**Pull request merged**